File: C/Algorithms/965.c
#include <stdbool.h>
#include <stdlib.h>

// Definition for a binary tree node.
struct TreeNode {
  int val;
  struct TreeNode *left;
  struct TreeNode *right;
};

bool isUnivalTree(struct TreeNode *node) {
  int size = 4, length = 0, value = node->val;
  struct TreeNode **stack = malloc(sizeof(*stack) * size);
  stack[length++] = node;
  while (length > 0) {
    node = stack[--length];
    struct TreeNode *left = node->left, *right = node->right;
    if (right) {
      if (right->val != value) {
        free(stack);
        return false;
      }
      if (length == size) {
        size *= 1.5;
        stack = realloc(stack, size * sizeof(*stack));
      }
      stack[length++] = right;
    }
    if (left) {
      if (left->val != value) {
        free(stack);
        return false;
      }
      if (length == size) {
        size *= 1.5;
        stack = realloc(stack, size * sizeof(*stack));
      }
      stack[length++] = left;
    }
  }
  free(stack);
  return true;
}

File: Rust/Algorithms/965.rs
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn is_unival_tree(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        let value = root.as_ref().unwrap().borrow().val;
        let mut stack = vec![root];

        while stack.len() > 0 {
            let node = Rc::try_unwrap(stack.pop().flatten().unwrap())
                .ok()
                .unwrap()
                .into_inner();
            let left = node.left;
            let right = node.right;

            if right.is_some() {
                if right.as_ref().unwrap().borrow().val != value {
                    return false;
                }
                stack.push(right);
            }

            if left.is_some() {
                if left.as_ref().unwrap().borrow().val != value {
                    return false;
                }
                stack.push(left);
            }
        }
        true
    }
}
